File: languages/python/04-expert/performance-and-profiling-basics/example/main.py
# Module focus: Measuring hot paths before changing code for speed.
# Why it matters: practicing performance and profiling basics patterns makes exercises and
# checkpoints easier to reason about.

from __future__ import annotations

import time

# Prepare sample inputs that exercise the key performance and profiling basics path.
retained_object: object | None = None


# Helper setup for performance and profiling basics; this keeps the walkthrough readable.
def measure_average(action, repetitions: int) -> float:
    # Warm up once so setup effects are less likely to dominate the repeated samples.
    action()

    total = 0.0
    for _ in range(repetitions):
        start = time.perf_counter()
        action()
        total += time.perf_counter() - start
    return total / repetitions


def build_with_concatenation(line_count: int) -> str:
    # Repeated concatenation creates new strings as the result grows.
    text = ""
    for index in range(line_count):
        text += f"row-{index};"
    return text


def build_with_join(line_count: int) -> str:
    # Building parts first lets join allocate the final string once.
    parts = [f"row-{index};" for index in range(line_count)]
    return "".join(parts)


def fill_without_presize(item_count: int) -> list[int]:
    values: list[int] = []
    for index in range(item_count):
        values.append(index)
    return values


def fill_with_presize(item_count: int) -> list[int]:
    # Pre-sizing makes the target length explicit before assignments begin.
    values = [0] * item_count
    for index in range(item_count):
        values[index] = index
    return values


# Walk through one fixed scenario so performance and profiling basics behavior stays repeatable.
def main() -> None:
    line_count = 4_000
    repetitions = 12

    def remember(value: object) -> None:
        # Keep results alive so the benchmark still performs real work.
        global retained_object
        retained_object = value

    # Measure two implementations of the same string-building task.
    concat_duration = measure_average(
        lambda: remember(build_with_concatenation(line_count)),
        repetitions,
    )
    join_duration = measure_average(
        lambda: remember(build_with_join(line_count)),
        repetitions,
    )

    # Report output values so learners can verify the performance and profiling basics outcome.
    print(f"Average string concatenation ({repetitions} runs): {concat_duration:.6f}s")
    print(f"Average str.join ({repetitions} runs): {join_duration:.6f}s")

    item_count = 200_000
    # Repeat the same measurement shape for list growth strategies.
    no_presize = measure_average(
        lambda: remember(fill_without_presize(item_count)),
        repetitions,
    )
    with_presize = measure_average(
        lambda: remember(fill_with_presize(item_count)),
        repetitions,
    )

    print(f"Average list fill without pre-size ({repetitions} runs): {no_presize:.6f}s")
    print(f"Average list fill with pre-size ({repetitions} runs): {with_presize:.6f}s")


if __name__ == "__main__":
    main()

File: languages/java/04-expert/performance-and-profiling-basics/example/Main.java
// Module focus: Measuring hot paths before changing code for speed.
// Why it matters: the example makes it possible to measure before optimizing and interpret timing
// data cautiously before the learner tackles the exercises.

import java.util.HashSet;
import java.util.Set;
import java.util.function.IntSupplier;
import java.util.stream.IntStream;

public class Main {
    private static final int MEASURED_RUNS = 8;
    // A volatile sink keeps every measured result observable outside the timed method.
    private static volatile int resultSink;

    public static void main(String[] args) {
        // Build both data structures before timing so setup is excluded from the comparison.
        int[] values = IntStream.range(0, 50_000).toArray();
        Set<Integer> indexedValues = new HashSet<>(values.length * 2);
        for (int value : values) {
            indexedValues.add(value);
        }

        int target = values[values.length - 1];
        IntSupplier linearWork = () -> repeatedLinearSearch(values, target, 100);
        IntSupplier indexedWork = () -> repeatedSetLookup(indexedValues, target, 100);

        // Warm both paths before measuring so first-use JVM work is less dominant.
        warmUp(linearWork);
        warmUp(indexedWork);

        long linearAverage = measureAverageNanos(linearWork);
        int linearMatches = resultSink;
        long indexedAverage = measureAverageNanos(indexedWork);
        int indexedMatches = resultSink;

        // Report output labels and correctness deterministically even though timings vary.
        System.out.println("Matches agree: " + (linearMatches == indexedMatches));
        System.out.println("Linear search average ns: " + linearAverage);
        System.out.println("HashSet lookup average ns: " + indexedAverage);
        System.out.println("Measured runs: " + MEASURED_RUNS);
    }

    private static int repeatedLinearSearch(int[] values, int target, int repetitions) {
        int matches = 0;
        // Each lookup may scan the full array, making the repeated cost intentionally visible.
        for (int attempt = 0; attempt < repetitions; attempt++) {
            for (int value : values) {
                if (value == target) {
                    matches++;
                    break;
                }
            }
        }
        return matches;
    }

    private static int repeatedSetLookup(Set<Integer> values, int target, int repetitions) {
        int matches = 0;
        // The set pays an indexing cost before measurement to provide direct lookup behavior.
        for (int attempt = 0; attempt < repetitions; attempt++) {
            if (values.contains(target)) {
                matches++;
            }
        }
        return matches;
    }

    private static void warmUp(IntSupplier action) {
        for (int iteration = 0; iteration < 3; iteration++) {
            resultSink = action.getAsInt();
        }
    }

    private static long measureAverageNanos(IntSupplier action) {
        long total = 0;
        // Averaging repeated samples reduces the influence of one scheduling interruption.
        for (int iteration = 0; iteration < MEASURED_RUNS; iteration++) {
            long started = System.nanoTime();
            resultSink = action.getAsInt();
            total += System.nanoTime() - started;
        }
        return total / MEASURED_RUNS;
    }
}
